WIP: my enhancements to enums

JeffBezanson opened the pull request:
I don't like to complain unproductively, so here I present Jeffcrosoft Enterprise Enums 2015, Server Edition.

Summary:
- faster, non-allocating constructor
- fast special case for contiguous values
- inner constructor only, for stronger abstraction
- O(1) method definitions per type
- supports enums with type parameters (weird, but why not)
- supports non-bits integer types
- `names` returns a tuple, which can hopefully be constant-folded in some cases

To do:
- testing
- I'm not sure `names` is the right function for getting a collection of all instances
- The previous version allowed "invalid" enums; not sure if that was intentional
- Fast special case for a small number of non-contiguous values

JeffBezanson committed:
my enhancements to enums

- faster, non-allocating constructor
- fast special case for contiguous values
- O(1) method definitions per type
- supports enums with type parameters (weird, but why not)

jakebolewski commented on base/Enums.jl:
```diff
@@ -67,32 +69,40 @@ macro enum(T,syms...)
+    vals = map(x->(x[1],convert(enumT,x[2])), vals)
+    quotednames = map(x->Meta.quot(x[1]), vals)
+    all_instances = map(x->Expr(:call, T, x[2]), vals)
+    valueset = Set(map(x->x[2], vals))
```
Could you do something like 
```julia
foldl((x1,x2)->Expr(:||, x1, Expr(:comparison, :x, :(==), x2)),
                          Expr(:comparison, :x, :(==), vals[1]), vals[2:end])
```
when the number of vals is small?

JeffBezanson replied:
Yep, that's the plan.

jakebolewski commented on base/Enums.jl:
```diff
@@ -47,19 +44,24 @@ macro enum(T,syms...)
         lo = min(lo, i)
         hi = max(hi, i)
+        if !first
+            contiguous &= (i == prev+1)
```
Another common idiom is Enum's which count down.  It would be nice to special case that as well.

JeffBezanson committed:
even more enum improvements:

- incorporate @quinnj's improvements
- add fast path for a small number of non-contiguous values
- handle contiguous values listed in any order
- add a `print` method that doesn't show the type

JeffBezanson committed:
more Enum tweaks

quinnj commented:
+1

StefanKarpinski commented:
How about holding off on supporting enums with type parameters until there's an actual use case to give us some sense of how it should work? Otherwise you know someone's going to use it and their code will break when we change this feature.

quinnj commented:
`Enums.jl` is certainly a fun exercise in meta-programming that isn't too crazy, but has some nice tricks. I might do a quick write up blog post; I think folks would generally find it interesting.

@StefanKarpinski, probably a good idea. I was originally thinking it would be useful to be able to write something like `@enum RedFruit<:Fruit apple cherry pomegranate` to which would translate to `abstract Fruit <: Enum` and `immutable RedFruit <: Fruit ...`. So you could then do `@enum YelloFruit<:Fruit banana lemon` to specify "groups" of enums.

Not sure on the use case for type parameters in enums.

JeffBezanson commented:
I agree the use cases for type parameters aren't clear, however I think it *is* pretty clear how they would have to work. Of course, if the behavior of an enum depends on its type parameters in various ways, I don't think the macro can be of any use; you have to implement everything yourself.

Being able to specify a supertype would be a good feature. But I don't think the enum macro should generate the supertype definition; it can just use the supertype you specify.

JeffBezanson commented:
To put it differently, supporting type parameters isn't really some fancy feature, it's just passing things through. It basically consists of defining `f{T<:MyType}(::Type{T})` instead of `f(::Type{MyType})`.

However if you want to turn it off the best course of action is just to insert an error at the top of the macro for a non-symbol argument. That way it's super easy to enable or disable.
